### ISSUE TITLE
[HOLD for payment 2023-04-27] [$1000] The attachment viewer loads every image again when switching with arrows

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#15922]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open any report with more than one image sent
2. Click on any one image
3. Click on an arrow to switch image

## Expected Result
The spinner should only display once when the image loads, after that, the image should not need to show the spinner and be shown instantly

## Actual Result
The spinner is shown on every image when switching and images are loaded again every time.

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [x] Android / Chrome
- [x] iOS / native
- [x] iOS / Safari
- [x] MacOS / Chrome / Safari
- [x] MacOS / Desktop

**Version Number:** 1.2.83-1
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/224829096-7835d17f-58a9-4d71-b04b-db2b30a79cdd.mov

https://user-images.githubusercontent.com/43996225/224829210-cfb2f08f-2543-4d2e-bcc3-66e5f90da9d6.mp4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @esh-g 
**Slack conversation:** 
https://expensify.slack.com/archives/C049HHMV9SM/p1678613459293019
[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0136002576e7090b5b</li>
        <li>Upwork Job ID: 1636029895829024768</li>
        <li>Last Price Increase: 2023-03-15</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @isabelastisser (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~0136002576e7090b5b

**Comment by MelvinBot** [comment]
Current assignee @isabelastisser is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @thesahindia (`External`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @johnmlee101 (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by isabelastisser** [comment]
sorry, forgot to check the boxes before applying the external label!

**Comment by johnmlee101** [comment]
I think this is a dupe of https://github.com/Expensify/App/issues/6527 which was deferred for https://github.com/Expensify/App/issues/12603 which ultimately deals with caching these images

**Comment by esh-g** [comment]
@johnmlee101 Are you sure this is a dupe of #6527 ?
Because I think that issue was even before the carousel control was implemented and was about caching the images to the browser so they stay even after reload. But here, the images are already loaded and we can have them switch without caching them to the browser.

**Comment by nefedov-dm** [comment]
**Proposal**

I think you need to render each `AttachmentView` for every attachment in `AttachmentCarousel` and use style `display: none;`. Because your AttachmentView will re-render every time when current slide will be change.

You can cache image content or make `Set` with loaded urls, but i think it more complicated and redundant for `Image` component.

Plus, if you will use style, you can use opacity and it is good for animation.

**Comment by MelvinBot** [comment]
📣 @nefedov-dm! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/202326034-1a292e64-48fd-4fef-9f4f-7feaaeedfd9f.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by usenbekov** [comment]
**Proposal**

The main reason is at the `Image` component used `debouncedConfigureImageSource` with 220 ms:
https://github.com/Expensify/App/blob/main/src/components/Image/index.js#L14

```
this.debouncedConfigureImageSource = _.debounce(this.configureImageSource, 220);
...
this.debouncedConfigureImageSource();
...
this.debouncedConfigureImageSource.cancel();
this.debouncedConfigureImageSource();
```

It will be fixed if we use it with the `immediate: true` option:
`this.debouncedConfigureImageSource = _.debounce(this.configureImageSource, 220, true);`

Or if we use it without `debounce`.

It needs more investigation like why the debounce actually used to make sure that the fix won't break other parts.

**Comment by usenbekov** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/15922#issuecomment-1471106265)

Updated according to contributing guidelines.

**Comment by Santhosh-Sellavel** [comment]
This is expected behavior, debounce here is added to ensure there is no flickering of the image while switching images fast.

cc: @chiragsalian @luacmartins @trjExpensify @thesahindia

**Comment by isabelastisser** [comment]
Not overdue, still discussing!

@johnmlee101 is this a dupe or expected behavior? Let me know if I should close it.

**Comment by usenbekov** [comment]
> This is expected behavior, debounce here is added to ensure there is no flickering of the image while switching images fast.
> 
> cc: @chiragsalian @luacmartins @trjExpensify @thesahindia

Image component mounted -> after 220ms -> configureImageSource -> RNImage.getSize -> onLoad -> isLoading: false

So even if the image is getting from the cache it waits for 220ms and then calls onLoad. If I'm not missing anything then I think it shouldn't be like that.

**Comment by Santhosh-Sellavel** [comment]
@usenbekov 
That's right, but it should be like that, and debounce duration is added to avoid unnecessary **flickering while switching between images faster.**

Moreover, we are planning to improve the `Image` to remove `debounce` for some cases and that's in planning we will get started on that next week. 

@chiragsalian or @luacmartins Can you comment on this thanks!

cc:  @thesahindia @johnmlee101

**Comment by kidroca** [comment]
Hey guys I think my PR here, improves the situation a bit  
- https://github.com/Expensify/App/pull/15663

It fixes a workaround we had with image flickering for avatars, by fixing `react-native-web` bug where internal image cache was ignored (delaying the displaying of an already loaded image a bit)

The PR also makes the `debounce` here unnecessary - we no longer have a flicker between prev/next image
https://github.com/Expensify/App/blob/0b24114bfd99f92e24f849e46a7ed2622adbee7c/src/components/Image/index.js#L14

Check this out:

https://user-images.githubusercontent.com/12156624/226054739-1c464e09-efb4-4a92-8064-737c5cb63f70.mov

_Yeah, the spinner still shows briefly, so I like the idea of waiting a bit before rendering it_ 
_(BTW this is DEV, so it's possible the spinner would not show at all on a PROD build)_

**Comment by kidroca** [comment]
The Carousel strategy might be improved a bit if we include image preloading 

1. When we open an attachment / carousel (modal)
2. An image is starts loading (we see a spinner then the image)
3. After the image loads we `Image.preload` the image before/after (e.g. the one that would be shown when we press an arrow)
4. Since `Image.preload` returns a promise we can track whether it's still loading or not (images from cache would preload instantly)
5. When the actual arrow is pressed - we only show loading if the image is still (pre)loading - otherwise we don't show a spinner but directly display the image

**Comment by kidroca** [comment]
> @kidroca I couldn't find `Image.preload` method in RN docs, is it your custom `method`?

Sorry the name is actually `prefetch`, it's this one https://reactnative.dev/docs/image#prefetch

**Comment by luacmartins** [comment]
@Santhosh-Sellavel WRT [this](https://github.com/Expensify/App/issues/15922#issuecomment-1474369238), yes we plan to improve the `Image` component by removing the debounce. Not sure if that will get rid of the loading issue though.

**Comment by Santhosh-Sellavel** [comment]
Yes, I agree!

**Comment by johnmlee101** [comment]
@Santhosh-Sellavel so will this be resolved in that issue?

**Comment by Santhosh-Sellavel** [comment]
It's best to hold until  #15663  is merged

**Comment by isabelastisser** [comment]
Not overdue, still on hold!

**Comment by isabelastisser** [comment]
Hey @johnmlee101 , this is merged: https://github.com/Expensify/App/pull/15663

Does it mean that we can choose a proposal for this issue and hire contributors in Upwork?

Thanks!

**Comment by kidroca** [comment]
Applied on Upwork
I'll post a PR by EOW

**Comment by isabelastisser** [comment]
Hired in Upwork:

@kidroca - approved proposal
@thesahindia - C+ review
@esh-g - reported the issue

**Comment by MelvinBot** [comment]
@johnmlee101 @kidroca @isabelastisser @thesahindia this issue is now 3 weeks old. There is one more week left before this issue <a href='https://stackoverflowteams.com/c/expensify/questions/15233/15234#15234'>breaks WAQ</a> and will need to go internal. What needs to happen to get a PR in review this week? Please create a thread in #expensify-open-source to discuss. Thanks!

**Comment by kidroca** [comment]
Still working on the PR

_The issue is a bit old, because it was on hold_

**Comment by kidroca** [comment]
Still working, should be ready tomorrow

A little teaser

https://user-images.githubusercontent.com/12156624/229907353-80d3e7ef-ef2e-45fa-95b1-15d779e0a9b8.mp4

**Comment by Santhosh-Sellavel** [comment]
Looks much better @kidroca!

**Comment by kidroca** [comment]
PR ready for review
- https://github.com/Expensify/App/pull/16973

**Comment by MelvinBot** [comment]
@johnmlee101, @kidroca, @isabelastisser, @thesahindia Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by kidroca** [comment]
PR is ready for review

- https://github.com/Expensify/App/pull/16973

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.3.2-5 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/16973


If no regressions arise, payment will be issued on 2023-04-27. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by MelvinBot** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@thesahindia] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@thesahindia] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@thesahindia] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@isabelastisser] Determine if we should create a regression test for this bug.
- [ ] [@thesahindia] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@isabelastisser] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by situchan** [comment]
There seems [regression](https://github.com/Expensify/App/issues/17723) from [PR](https://github.com/Expensify/App/pull/16973) which fixes this issue

**Comment by trjExpensify** [comment]
@situchan can you clarify what you mean. Are you saying the PR for this issue caused a regression?

**Comment by situchan** [comment]
> @situchan can you clarify what you mean. Are you saying the PR for this issue caused a regression?

yes

**Comment by isabelastisser** [comment]
To confirm:

@kidroca: approved proposal 
@thesahindia - C+ review
@esh-g - reported the issue

-Contributor assigned on March 29. 
- PR merged on April 19. 
- Deployed to production on April 20.

The issue amount is $1000, but does it mean that the payment processed will be $500 because: `Merged PR more than 9 business days after assignment - 50% penalty`

@johnmlee101 ?

**Comment by isabelastisser** [comment]
Not overdue, still discussing this with @johnmlee101 !

**Comment by johnmlee101** [comment]
@isabelastisser This was a fairly complex issue that had a lot of edge cases that were being tested, so I think the 50% penalty might not apply in this case. I'd be fine keeping it as the original $1000 for this.

**Comment by isabelastisser** [comment]
Update: Payments made in Upwork and contracts ended.

@thesahindia, please complete your tasks on the `BugZero Checklist:`, and I will review and complete mine so we can close the issue. Thanks!

**Comment by isabelastisser** [comment]
QA issue [here.](https://github.com/Expensify/Expensify/issues/281199)

All set!
